### ISSUE TITLE
[TF] Help with TensorFlow version of Nadaraya-Watson Kernel Regression

### ISSUE THREAD

**Issue by floriandonhauser** [d2l-ai/d2l-en#1696]
I have finished a TensorFlow implementation of Nadaraya-Watson Kernel Regression.
Unfortunately, it might be numerically unstable or I still have a bug (the likely problem).
I am unable to find the problem :(
I have saved all random values and I load the from file to eliminate this change.
I initialize the learnable parameter w with the same constant and after training the constant is identical.
Unfortunately, the output of the model is different, very noisy for the TensorFlow version.
Attached, you can find my two numpy files as well as the versions in PyTorch and TensorFlow for testing.
Both load the values for x_train and y_train from the files.
PyTorch version:
![torch-nadaraya-watson](https://user-images.githubusercontent.com/35968719/112510565-d4821000-8d91-11eb-882d-2bc54f884d64.png)
TensorFlow version:
![tf-nadaraya-watson](https://user-images.githubusercontent.com/35968719/112510606-dd72e180-8d91-11eb-91f8-5a7931ce94fa.png)

[nadaraya-watson.zip](https://github.com/d2l-ai/d2l-en/files/6206374/nadaraya-watson.zip)

@terrytangyuan Maybe you have some insights or experience with similar problems.

**Comment by astonzhang** [comment]
Thanks. Can you check the default init methods for parameters?

**Comment by floriandonhauser** [comment]
In this example we initialize the single w number with a value from a uniform distribution [0, 1).
To make both implementations identical, I use the constant value 0.5.

**Comment by terrytangyuan** [comment]
Did you use high-level or low-level APIs to construct the experiment?

**Comment by floriandonhauser** [comment]
> Did you use high-level or low-level APIs to construct the experiment?

I am using tf.keras but I attached the code of my experiment above.
Maybe you can spot the problem.

**Comment by nikisix** [comment]
Looks like high-variance, did you remember your normalizing constant (dividing by 2) in the tf-kernel?

**Comment by floriandonhauser** [comment]
@nikisix 
I think I have the normalizing constant and the training seems to work.
w converges to the same value as far as I can tell.
The important parts that might contain the bug are:
```
class NWKernelRegression(tf.keras.Model):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.w = tf.Variable(initial_value=[0.5], trainable=True)

    def call(self, queries, keys, values):
        # Shape of the output `queries` and `attention_weights`:
        # (no. of queries, no. of key-value pairs)
        queries = tf.reshape(tf.repeat(queries, keys.shape[1]),
                             (-1, keys.shape[1]))
        self.attention_weights = tf.nn.softmax(
            -((queries - keys) * self.w)**2 / 2, axis=1)
        # Shape of `values`: (no. of queries, no. of key-value pairs)
        return tf.reshape(tf.matmul(tf.expand_dims(self.attention_weights, 1),
                                    tf.expand_dims(values, -1)), -1)
```
```
# Shape of `X_tile`: (`n_train`, `n_train`), where each column contains the
# same training inputs
X_tile = tf.reshape(tf.tile(x_train, [n_train]), [n_train, n_train])
# Shape of `Y_tile`: (`n_train`, `n_train`), where each column contains the
# same training outputs
Y_tile = tf.reshape(tf.tile(y_train, [n_train]), [n_train, n_train])
# Shape of `keys`: ('n_train', 'n_train' - 1)
keys = tf.reshape(X_tile[tf.math.logical_not(tf.eye(n_train, dtype=tf.bool))],
                  (n_train, -1))
# Shape of `values`: ('n_train', 'n_train' - 1)
values = tf.reshape(Y_tile[tf.math.logical_not(tf.eye(n_train, dtype=tf.bool))],
                  (n_train, -1))
```
```
net = NWKernelRegression()
trainer = tf.keras.optimizers.SGD(learning_rate=0.5)
animator = d2l.Animator(xlabel='epoch', ylabel='loss', xlim=[1, 5])

for epoch in range(5):
    with tf.GradientTape() as tape:
        # Note: L2 Loss = 1/2 * MSE Loss. Hence we halve the loss
        l = tf.reduce_sum(tf.square(net(x_train, keys, values) - y_train) / 2)
    grads = tape.gradient(l, net.trainable_weights)
    trainer.apply_gradients(zip(grads, net.trainable_weights))
    print(f'epoch {epoch + 1}, loss {float(l):.6f}')
    animator.add(epoch + 1, float(l))
```
```
# Shape of `keys`: (`n_test`, `n_train`), where each column contains the same
# training inputs (i.e., same keys)
keys = tf.reshape(tf.tile(x_train, [n_test]), [n_train, n_train])
# Shape of `value`: (`n_test`, `n_train`)
values = tf.reshape(tf.repeat(y_train, [n_test]), (n_test, n_test))
y_hat = tf.stop_gradient(tf.expand_dims(net(x_test, keys, values), 1))
plot_kernel_reg(y_hat)
```

**Comment by floriandonhauser** [comment]
@terrytangyuan @astonzhang 
If you want, I can also create a [WIP] Pull request so you have an easier way to look at the code and its output.
The code runs and the only problem is the instability of the model.

**Comment by nikisix** [comment]
Actually Florian, yours might best fit in the discussion section if I'm not mistaken.

**Comment by nikisix** [comment]
@floriandonhauser your formulas check out from what I can tell! Probably something else going on..

**Comment by biswajitsahoo1111** [comment]
@floriandonhauser please check this out: https://github.com/biswajitsahoo1111/D2L_Attention_Mechanisms_in_TF/blob/master/10_2_Attention_based_regression.ipynb

The plots are not identical to that of PyTorch's but similar, in my opinion. I have used default weight initializations of tf.keras.

**Comment by floriandonhauser** [comment]
@biswajitsahoo1111 That's great :)
Would you be able to create Pull Requests for the chapters to include them in the normal D2l website/book?
If you want, I can also help with that but I do not want to "steal" your code for the "d2l-en" repo and would prefer if you would be the contributor :)

**Comment by biswajitsahoo1111** [comment]
@floriandonhauser That's so nice of you. Let me try to create a pull request first. In case of any issues, I will definitely seek your help. Thanks.
